### ISSUE TITLE
Is Parrallel Decoding Possible?

### ISSUE THREAD

**Issue by NetaPanda** [ros-event-camera/event_camera_codecs#20]
Since the decoder keeps track of some time stamp information, the event msg should be decoded one by one. However, when the number of events is huge (like 180+MB/s), or the CPU is not strong (such as the one in an edge device), a single decoder thread can not satisfy real-time decoding and multi-thread parallel decoding is needed. I have tried initializing the decoder factory, Accumulator (copied from event_camera_py) and decoder in a new thread whenever a new event_camera_msg comes, but of course it gives the wrong event and trigger time stamps. I wonder is there any way to do parallel decoding? Maybe some code modification of the decoder is necessary? (I use EVK4 camera so I guess I need to modify the evt3 encoder)

**Comment by berndpfrommer** [comment]
I thought about it a bit but it will be hard work to parallelize the decoder.
Obviously the EVT3 format is not under our control since it comes like that raw from the camera (I saw that metavision 5 introduced EVT4 but I have not looked into that yet). So what options do we have?

As you pointed out the issue is with the incremental time stamps, and in particular the rollover of the TIME_HIGH field about every 16 seconds. But I'm not sure it ends there. What about the vector updates? Are we guaranteed that the SDK will deliver the VECT_BASE_X update and the subsequent VECT_12 update in the same packet? The ROS driver will not split them since it preserves SDK packet boundaries.

One could deal with the time stamps by having a single thread decode enough to determine the correct time of the first event of a packet, then dispatch the packet to a different pool of threads for complete decoding.
To find the correct starting time, one would have to decode backwards, starting at the tail of the previous packet, until a valid (TIME_HIGH/TIME_LOW) pair is found, then handle any potential TIME_HIGH rollovers (assuming that TIME_HIGH rolls over at most once inside a packet, which is reasonable). So it's doable, but will take a while to get right, and even longer to really see a good speedup.

And there is another reason why I haven't worked on this yet: on my 3-year old AMD laptop I get north of 100MeV/s decoding. Any subsequent algorithm I write bogs down well before then, making it mood to optimize the decoder. What application do you have that can handle such large event rates?

**Comment by berndpfrommer** [comment]
Just checked again: I'm also just getting 50-60 Mev/s on my laptop. Not sure what the problem is because I remember much higher numbers...

**Comment by NetaPanda** [comment]
Thanks for the detailed reply. 180+MB/s happens in extreme situations, such as under flickering 50-60Hz indoor lights. I use a Core-i9 13900K CPU and single thread works well around 70MB/s but not at 180+MB/s. I got these bandwidths using the metavision_driver node (the unit is MB/s instead of Mev/s). For a CPU like I9, I plan to achieve real-time decoding for arbitrary event rates. This can avoid a stuck ROS2 node when, for example, someone turns on a flickering light. More importantly, after testing on i9, I plan to migrate the ROS2 nodes to edge devices like NVIDIA Jetson. I guess Jetson can not do real-time decoding even on 50-60M  with its single 1.7GHz Arm core.

Since you mentioned that AMD laptop can do 100MeV/s decoding (which is much higher than 180MB/s considering one ev=few bytes), I suspect the issue comes from the ROS2 node I write. I have a subscription to the EventPacket msg, and in the callback function, I decode the msg and save events. Maybe I should just save the msg and decode them with another thread?

Based on the analysis you provided, it seems to be very complex to implement multi-thread decoding. I'll just accept the fact and try updating my code and using some event filters to reduce the event rate. Thanks for the help.

**Comment by NetaPanda** [comment]
Have tested saving EventPacket msgs in one thread and decoding with another standalone thread. The lag still exists at 180+MB/s.  The vector used to store EventPacket msgs has a growing size, which indicates slower decoding than real-time. Looks like the decoding speed of i9 is slower than that of the AMD cpu?

**Comment by NetaPanda** [comment]
The rate is around 150Mev/s (which I think it should be real-time on i9). Would a composable node save memory access time?  Currently I build a standalone node and run it, some pieces of code are listed below. I have limited experiments with ROS2, so it would be nice if this is just a coding issue and can be solved on my side.

Below is my callback function for the EventPacket subscriber:
```
   //only recieve, no decode
    void listener_callback(const event_camera_codecs::EventPacketConstSharedPtr & msg) {
      event_msg_list.push_back(std::make_shared<EventPacket>(*msg));
      return;
    }
```
And below is the function to decode them, which is a timer callback (runs with 0ms duration):

```
    void decoder_timer_callback()
    {
      event_camera_codecs::EventPacketConstSharedPtr msg;
      msg = event_msg_list.front();  // here the msg is already in the memory of this node
      event_msg_list.pop_front();
      std::cout<< "msg_list size" << event_msg_list.size() << std::endl;
      //return; // if we return here and not decode at all, msg_list size will always be 0
      DecodeEventMsg(msg); // would a composable node help?
      return;  // return here and the msg_list size will consistently increasing
     // rest codes are omitted
    }
```

This is the DecodeEventMsg function:
```
    void DecodeEventMsg(const event_camera_codecs::EventPacketConstSharedPtr & msg) {
      // will create a new decoder on first call, from then on returns existing one
      std::string encoding = msg->encoding;
      uint64_t timeBase = msg->time_base;
      auto decoder = decoderFactory.getInstance(*msg);
      if (!decoder) { // msg->encoding was invalid
        return;
      }
      decoder->setTimeMultiplier(1);
      decoder->setTimeBase(timeBase);
      // the decode() will trigger callbacks to processor
      processor.setHasSensorTimeSinceEpoch(decoder->hasSensorTimeSinceEpoch());
      processor.reset_stored_events();
      decoder->decode(*msg, &processor);
    }
```

**Comment by berndpfrommer** [comment]
The question is what your processor class does in its event handling method. Does it perform any heavy memory access?

Writing a composable ROS2 node is not hard, and it will avoid a copy of the undecoded event packet from the driver to your node.  Reducing memory traffic is what it's all about, and it makes performance debugging hard. Accessing memory in one node will all of a sudden slow down your code in another node. Or just subscribing to a topic will slow down another subscriber, because now there are two memory copies necessary. It can be very confusing.

For instance I have found that having a thread to decode the events, and then another thread to process them is *slower* than processing them single threaded (unless you have a very heavy processing step). This is because the decoded event packets are bloated (to e.g. 8 bytes/event), and the memory copies invalidate the processor cache, causing everything else to slow down. This could be better with a large processor cache, but for my AMD with its 4MB L2 / 8MB L3 cache this was a real issue.

This is also why event_camera_codecs does not offer a full-packet decode: just copying around the decoded packet incurs a big performance hit. Better to do the work right when the event is decoded.

**Comment by NetaPanda** [comment]
I appreciate the detailed suggestion! The processor is an `Accumulator` object, which I copied from [event_camera_py/accumulator.h](https://github.com/ros-event-camera/event_camera_py/blob/master/include/event_camera_py/accumulator.h). I just removed the pybind11 related parts for usage in c++. The code is pasted below.

Looks like it is not easy to find a quick answer. I will try implementing a composable ROS2 node and also try other solutions, especially on reducing memory traffic. If anything works, I will report here for reference. Thanks again for all the detailed explanations :)

```
class Accumulator : public event_camera_codecs::EventProcessor
{
public:
  // inherited from EventProcessor
  void eventCD(uint64_t sensor_time, uint16_t ex, uint16_t ey, uint8_t polarity) override
  {
    cdEvents_->push_back(EventCD(ex, ey, polarity, shorten_time(sensor_time)));
    maxSizeCD_ = std::max(cdEvents_->size(), maxSizeCD_);
    numCDEvents_[std::min(polarity, uint8_t(1))]++;
  }

  void eventExtTrigger(uint64_t sensor_time, uint8_t edge, uint8_t id) override
  {
    extTrigEvents_->push_back(EventExtTrig(
      static_cast<int16_t>(edge), static_cast<int64_t>(sensor_time), static_cast<int16_t>(id)));
    maxSizeExtTrig_ = std::max(extTrigEvents_->size(), maxSizeExtTrig_);
    numExtTrigEvents_[std::min(edge, uint8_t(1))]++;
  }

  void finished() override {}
  void rawData(const char *, size_t) override {}

  // own methods
  void reset_stored_events()
  {
    delete cdEvents_;  // in case events have not been picked up
    cdEvents_ = new std::vector<EventCD>();
    delete extTrigEvents_;  // in case events have not been picked up
    extTrigEvents_ = new std::vector<EventExtTrig>();
    // TODO(Bernd): use hack here to avoid initializing the memory
    cdEvents_->reserve(maxSizeCD_);
    extTrigEvents_->reserve(maxSizeExtTrig_);
  }
  uint64_t get_start_time() const { return (startTime_); }

std::vector<EventCD> * get_cd_events()
  {
    if (cdEvents_) {
      auto p = cdEvents_;
      return p;
    }
    return 0;
  }
 std::vector<EventExtTrig>* get_ext_trig_events()
  {
    if (extTrigEvents_) {
      auto p = extTrigEvents_;
      return p;
    }
    return 0;
  }
  ~Accumulator()
  {
    delete cdEvents_;
    delete extTrigEvents_;
  }
  size_t get_num_cd_off() const { return (numCDEvents_[0]); }
  size_t get_num_cd_on() const { return (numCDEvents_[1]); }
  size_t get_num_trigger_rising() const { return (numExtTrigEvents_[0]); }
  size_t get_num_trigger_falling() const { return (numExtTrigEvents_[1]); }
  void initialize(uint32_t, uint32_t) {}
  void setHasSensorTimeSinceEpoch(bool b) { hasSensorTimeSinceEpoch_ = b; }
  int32_t shorten_time(uint64_t t)
  {
    if (hasSensorTimeSinceEpoch_) {
      if (!hasStartTime_) {
        startTime_ = t;
        hasStartTime_ = true;
      }
      return (static_cast<int32_t>(t - startTime_));
    }
    return (static_cast<int32_t>(t - startTime_));
  }

private:
  // ------------ variables
  bool hasStartTime_{false};
  bool hasSensorTimeSinceEpoch_{false};
  uint64_t startTime_{0};
  size_t numCDEvents_[2] = {0, 0};
  size_t numExtTrigEvents_[2] = {0, 0};
  std::vector<EventCD> * cdEvents_{0};
  std::vector<EventExtTrig> * extTrigEvents_{0};
  size_t maxSizeCD_{0};
  size_t maxSizeExtTrig_{0};
};
```

**Comment by NetaPanda** [comment]
My latest test shows that, even if I do nothing and immediately return in the `eventCD(uint64_t sensor_time, uint16_t ex, uint16_t ey, uint8_t polarity) override` function in Accumulator (processor), there are still lags. That is to say, under the heavy event rate, simply passing through all events with each event triggers a callback function call, have already been slower than real-time. I will check out why this is the case.

**Comment by berndpfrommer** [comment]
Thanks for the feedback.
Removing the memory copies in the accumulator should have definitely improved the situation. However, the EVK4 sensor will be able

**Comment by NetaPanda** [comment]
I have tried a few methods here and there. I still can not get a concrete answer, but here are some facts:

1. Under flickering light and rapid movement, perf.cpp gives me the following:
```
number of events read: 1506.71 Mev in 14.9633 seconds
event rate: 100.694 Mevs
```
This means real-time decoding is impossible. I have also tried  [metavision's standalone EVT3 decoder](https://docs.prophesee.ai/stable/samples/standalone/evt3_decoder.html). If I save events to a fixed size vector (with idx % vec.size() to avoid overflow), that decoder gives the following:

```
Decoded '/home/gmy/tmp/test.raw' in 9.8052 s
Decoded events:  1506712242
Decoding Rates:  153.665 Mevs
```
Which is about real-time. I believe the event_camera_codecs sacrificed decoding speed in exchange for its generalization ability to many types of event encodings. But even this explicit EVT3 decoder is still not fast enough since I need to do not only saving but event processing. There is yet another decoder [here](https://docs.prophesee.ai/stable/data/streaming_decoding/sdk_api.html), I have not tested it since it has been integrated in metavision HAL sdk. **BUT** it is worth trying, since I use metavision studio to record test.raw, and it can display the event frames in real-time without lag **even on my old slow PC.** To get the event frame one has to decode the events, right? So I think metavision SDK can do event decoding really efficiently.

BTW: perf.cpp needs a few modifications. First, it needs a 
`decoder->setGeometry(1280, 720);`, otherwise the events are not accessed. Second, `sizeof(buffer)` in Line 125 is the size of the vector object, which is only 24 Bytes.

2. Speak of memory optimization, I found using a static global event buffer in the callback function is faster than creating a temporal vector inside it:
```
std::vector<EventCD> cdEventsVec(1000000);
uint32_t counter = 0;
class Accumulator : public event_camera_codecs::EventProcessor
{
public:
  // inherited from EventProcessor
  void eventCD(uint64_t sensor_time, uint16_t ex, uint16_t ey, uint8_t polarity) override
  {
    if (counter >= cdEventsVec.size()) {
      RCLCPP_INFO(rclcpp::get_logger("Accumulator"), "Resize cdEventsVec");
      cdEventsVec.resize(cdEventsVec.size() + 100000);
    }
    cdEventsVec[counter].x = ex;
    cdEventsVec[counter].y = ey;
    cdEventsVec[counter].p = polarity;
    cdEventsVec[counter++].t = static_cast<int32_t>(sensor_time);
  }
```
This may be caused by the increased cache hit rate.

Anyway, what I want to achieve is decoding and storing the events, then publishing them with ROS. I use another node to receive these events and convert them to voxels. I would like these operations to be real-timed. Based on our conversation, I feel this might be an impossible task, unless ERC is applied. My best practice now supports only 50Mev/s (100 MB/s), and strange lag can still happen once the nodes runs for a few seconds.
